src/source-control/GitSourceControl.ts

```diff
@@ -1,42 +1,396 @@
+              return newCommitOid.tostrS();
+            }
+            return parent;
+          });
+          childCommit.parentCommits = parentsOfChild;
+          draftCommitHashMap.set(child, childCommit);
+        });
+      });
+    }
+    // Remove temp branch
+    nodegit.Branch.delete(await repo.getBranch(tempBranchName));
+    // SEE: Call loadRepository? How?
   }
 
   async pushCommit(commit: Commit): Promise<void> {
-    // TODO: Implement
+    try {
+      const {
+        userPublicKeyPath,
+        userPrivateKeyPath,
+        userPassphrase,
+      } = JSON.parse(
+        fs.readFileSync(`${this.repoPath}/sttack.config.json`, "utf-8"),
+      );
+      const repo = await nodegit.Repository.open(this.repoPath);
+      // SEE: Could be any of the long lived branches?
+      const remote = await repo.getRemote("origin");
+      const callback = {
+        credentials: function (_url: string, userName: string) {
+          //console.log("Getting credentials");
+          //   // FIXME: Possible infinite loop when using sshKeyFromAgent
+          //   // See: https://github.com/nodegit/nodegit/issues/1133
+          //   //return nodegit.Cred.sshKeyFromAgent(userName);
+          return nodegit.Cred.sshKeyNew(
+            userName,
+            userPublicKeyPath, //"/Users/phuonganh/.ssh/id_rsa.pub",
+            userPrivateKeyPath, //"/Users/phuonganh/.ssh/id_rsa",
+            userPassphrase, //"hello",
+          );
+        },
+      };
+      const connection = remote.connect(nodegit.Enums.DIRECTION.PUSH, callback);
+      // SEE: Can only push if branch exists on remote? How to find branch name here?
+      await Promise.all(
+        commit.refNames.map(
+          async (ref: string): Promise<number> => {
+            return await remote.push([`${ref}:${ref}`], {
+              callbacks: {
+                credentials: function (_: string, userName: string) {
+                  return nodegit.Cred.sshKeyNew(
+                    userName,
+                    userPublicKeyPath, //"/Users/phuonganh/.ssh/id_rsa.pub",
+                    userPrivateKeyPath, //"/Users/phuonganh/.ssh/id_rsa",
+                    userPassphrase, //"hello",
+                  );
+                },
+              },
+            });
+          },
+        ),
+      );
+    } catch (err) {
+      console.log(err);
+    }
   }
 
   async pushCommitsForCommitTreeRootedAtCommit(commit: Commit): Promise<void> {
-    // TODO: Implement
+    const queue: Commit[] = [commit];
+    while (queue.length) {
+      // SEE: Always have an element there
+      const commit = queue.pop()!;
+      await this.pushCommit(commit);
+      queue.push(
+        ...commit.childCommits.map((hash) => this.commitHashMap.get(hash)!),
+      );
+    }
+    // SEE: Call Load Repository Information
   }
 
   async attachSttackBranchesToCommits(commits: Commit[]): Promise<void> {
-    // TODO: Implement
+    // TODO: Unsure
```

Comment:
We can use @manyaagarwal's implementation in #30 here, but we need to undo this change to prevent merge conflicts

```suggestion
    // TODO: Implement
```

Reply:
Ya, I have shifted the branch name code here. The branch creation would still have to be implemented. Pushing my changes soon. Shifting stuff around.

Reply:
Okay #30 doesn't implement this. We'll need to generate branch names (which #30 does), then do the `git branch -f` equivalent that our existing code is already doing. I can find this out for you if you need; we already have the code to do this.

#30 will need a different method signature though, so you'll need to sync with @manyaagarwal